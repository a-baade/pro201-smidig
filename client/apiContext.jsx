import React from "react";
import { fetchJSON } from "./lib/fetchJSON";
import { postJSON } from "./lib/postJSON";

export const ApiContext = React.createContext({
  async fetchLogin() {
    return await fetchJSON("/api/login");
  },

  async allCharities(query) {
    return await fetchJSON("/api/charities?" + new URLSearchParams(query));
  },

  async oneCharity(id) {
    return await fetchJSON("/api/charities/charity/id?id=" + id);
  },

  async donateCharity(id) {
    return await fetchJSON("/api/charities/donate/id?id=" + id);
  },

  async registerLogin(provider, login) {
    return await postJSON(`/api/login/${provider}`, login);
  },

  async endSession() {
    const res = await fetch("/api/login", { method: "DELETE" });
    if (!res.ok) {
      throw new Error(`Failed to post ${res.status}: ${res.statusText}`);
    }
  },
});
