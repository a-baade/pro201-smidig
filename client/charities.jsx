import React from "react";

export function Charities() {
  return <h1>All Charities</h1>;
}