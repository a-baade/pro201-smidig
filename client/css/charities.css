/* style for charities.jsx */

.charity-cards-container{
    display: flex;
    flex-direction: row;
    justify-content: space-evenly;
    flex-wrap: wrap;
    width: 100%;
    padding: 0;
}

.charity-card {
    max-width: 35vw;
    min-width: 450px;
    margin-top: 10vh;
    line-height: 25px;
    font-size: 20px;
    text-align: center;
    box-sizing: border-box;
    /*border: 2px solid #ACCCCE;*/
    border-top-left-radius: 20px;
    border-top-right-radius: 20px;
}

.charity-background {
    border-top-left-radius: 19px;
    border-top-right-radius: 19px;
    background-size: cover;
    background-position: center;
    background-repeat: no-repeat;
    height: 200px;
}

.charity-img {
    position: relative;
    top: 120px;
    height: auto;
    width: 170px;
    border-radius: 50%;
    object-fit: cover;
}

.charity-card-content {
    margin-top: 65px;
    padding: 20px;
}

.card-buttons {
    display: flex;
    flex-direction: row;
    justify-content: space-evenly;
}

.card-button {
    background: #ACCCCE;
    border-radius: 10px;
    border: 0;
    text-transform: uppercase;
    letter-spacing: 1px;
    font-size: 12px;
    font-weight: 600;
    margin: 10px 25px;
    padding: 17px 32px;
    width: 170px;
}


.search-charities {
    margin-left: 30px;
    margin-top: 30px;
    width: 100%;
    position: relative;
    display: flex;
    box-shadow: -15px 15px 17px -10px rgba(0,10,10,0);
}
.search-input-charities{
    width: 100%;
    padding: 9px;
    border-radius:10px 10px 1px 10px;

    outline: none;
    font-size: 20px;
    background: none;
}
.charity-results {
    position: absolute;
    padding-top: 63px ;
    padding-left: 23px;
    text-decoration: underline;
    text-underline-offset: 3px;
}
.container-charities{
    width: 35%;
    left: 40%;
    top: 40%;
}

/* ? */
.pic-container {
    position: relative;
    z-index: -1;
}