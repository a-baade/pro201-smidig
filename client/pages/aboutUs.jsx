import React from "react";
import MainBg from "../pics/about-us-pictures/main-bg.png";
import FirstLayerArt from "../pics/about-us-pictures/firstLayerArt.png";
import TopLayerArt from "../pics/about-us-pictures/topLayerArt.png";
import PovertyFirst from "../pics/about-us-pictures/povertyFirstLayer.png";
import PovertySecond from "../pics/about-us-pictures/povertytopLayer.png";
import ZeroHunger from "../pics/about-us-pictures/zeroHungerFirst.png";
import ZeroHungerTop from "../pics/about-us-pictures/seroHungerTop.png";
import GoodHealthFirst from "../pics/about-us-pictures/goodHealthFirst.png";
import GoodHealthTop from "../pics/about-us-pictures/goodHealthTop.png";
import QualityEducation from "../pics/about-us-pictures/qualityEducation.png";
import QualityEducationTop from "../pics/about-us-pictures/qualityEducationTop.png";
import GenderEquality from "../pics/about-us-pictures/genderEqualityFirst.png";
import GenderEqualityTop from "../pics/about-us-pictures/genderEqualityTop.png";

export function AboutUs() {
  return (
    <div>
      <div>
        <img src={MainBg} className={"main-bg-pic"} alt={"mainBg"}/>
      </div>
      <div className={"second-row-box"}>
        <div className={"knowledge-text-box"}>
          <p className={"our-mission"}>OUR MISSION</p>
          <h1 className={"knowledge-power"}>KNOWLEDGE = POWER</h1>
          <p>We're a Norwegian female-founded startup changing corporate giving and the way companies and nonprofits connect.</p>
          <p>We're saving companies resources, help create stable income streams for nonprofits, and create deeper relationships between donor and nonprofit.</p>
          <p> We help implement solid strategies for impactful and sustainable giving, all in line with the UN SDGs.</p>
        </div>
        <div className={"art-box"}>
          <img src={FirstLayerArt} className={"first-layer-art"} alt={"firstLayer"}/>
          <img src={TopLayerArt} className={"top-layer-art"} alt={"topLayer"}/>
          <p className={"our-mission mission"}>OUR MISSION</p>
          <h1 className={"action-call"}>A CALL TO ACTION</h1>
        </div>
      </div>
      <div className={"big-container"}>
        <div className={"small-box light"}>
          <p className={"small-title"}>No poverty</p>
          <p className={"small-txt"}>We are committed to end poverty in all its forms everywhere</p>
          <img src={PovertyFirst} className={"poverty-first"}/>
          <img src={PovertySecond} className={"poverty-second"}/>
        </div>
        <div className={"small-box dark"}>
          <p className={"small-title"}>Zero hunger</p>
          <p className={"small-txt"}>End hunger, achieve food security and improve nutrition and promote sustainable agriculture</p>
          <img src={ZeroHunger} className={"zero-hunger"}/>
          <img src={ZeroHungerTop} className={"zero-hunger-top"}/>
        </div>
        <div className={"small-box light"}>
          <img src={GoodHealthFirst} className={"good-health"}/>
          <img src={GoodHealthTop} className={"good-health-top"}/>
          <p className={"small-title good-health-title"}>Good health</p>
          <p className={"small-txt good-health-txt"}>Ensure healthy lives and promote well-being for all at all ages</p>
        </div>
        <div className={"small-box light"}>
          <p className={"small-title"}>Quality education</p>
          <p className={"small-txt"}>Ensure inclusive and equitable quality education and promote lifelong learning opportunities for all</p>
          <img src={QualityEducation} className={"quality-education"}/>
          <img src={QualityEducationTop} className={"quality-education-top"}/>
        </div>
        <div className={"small-box dark"}>
          <img src={GenderEquality} className={"gender-equality"}/>
          <img src={GenderEqualityTop} className={"gender-equality-top"}/>
          <p className={"small-title gender-eql-title"}>Gender equality</p>
          <p className={"small-txt gender-eql-txt"}>Achieve gender equality and empower all women and girls</p>
        </div>
        <div className={"small-box light"}>
          <p className={"small-title"}>Clean water & sanitation</p>
          <p className={"small-txt"}>Ensure availability and sustainable management of water and sanitation for all</p>
          <img src={ZeroHunger} className={"zero-hunger"}/>
          <img src={ZeroHungerTop} className={"zero-hunger-top"}/>
        </div>
        <div className={"small-box light"}>
          <p className={"small-title"}>Affordable & clean energy</p>
          <p className={"small-txt"}>Ensure access to affordable, reliable, sustainable and modern energy for all</p>
          <img src={PovertyFirst} className={"poverty-first"}/>
          <img src={PovertySecond} className={"poverty-second"}/>
        </div>
        <div className={"small-box light"}>
          <p className={"small-title"}>Decent work & economic growth</p>
          <p className={"small-txt"}>Promote sustained, inclusive and sustainable economic growth, full and productive employment and decent work for all</p>
          <img src={ZeroHunger} className={"decent-work"}/>
          <img src={ZeroHungerTop} className={"decent-work-top"}/>
        </div>
        <div className={"small-box light"}>
          <p className={"small-title"}>Industry, innovation & infrastructure</p>
          <p className={"small-txt"}>Build resilient infrastructure, promote inclusive and sustainable industrialization and foster innovation</p>
          <img src={PovertyFirst} className={"industry"}/>
          <img src={PovertySecond} className={"industry-second"}/>
        </div>
        <div className={"small-box dark"}>
          <p className={"small-title"}>Reduce inequalities</p>
          <p className={"small-txt"}>Reduce inequality within and among countries</p>
          <img src={ZeroHunger} className={"zero-hunger"}/>
          <img src={ZeroHungerTop} className={"zero-hunger-top"}/>
        </div>
        <div className={"small-box light"}>
          <p className={"small-title"}>Subsainable cities & communinites</p>
          <p className={"small-txt"}>Make cities and human settlements inclusive, safe, resilient and sustainable</p>
          <img src={PovertyFirst} className={"poverty-first"}/>
          <img src={PovertySecond} className={"poverty-second"}/>
        </div>
        <div className={"small-box dark"}>
          <p className={"small-title"}>responsible consumption & production</p>
          <p className={"small-txt"}>Ensure sustainable consumption and production patterns</p>
          <img src={ZeroHunger} className={"zero-hunger"}/>
          <img src={ZeroHungerTop} className={"zero-hunger-top"}/>
        </div>
        <div className={"small-box light"}>
          <p className={"small-title"}>Climate action</p>
          <p className={"small-txt"}>Take urgent action to combat climate change and its impacts</p>
          <img src={PovertyFirst} className={"poverty-first"}/>
          <img src={PovertySecond} className={"poverty-second"}/>
        </div>
        <div className={"small-box dark"}>
          <p className={"small-title"}>Life below water</p>
          <p className={"small-txt"}>Conserve and sustainably use the oceans, seas and marine resources for sustainable development</p>
          <img src={ZeroHunger} className={"zero-hunger"}/>
          <img src={ZeroHungerTop} className={"zero-hunger-top"}/>
        </div>
        <div className={"small-box light"}>
          <p className={"small-title"}>Life on land</p>
          <p className={"small-txt"}>Protect, restore and promote sustainable use of terrestrial ecosystems, sustainably manage forests, combat desertification, and halt and reverse land degradation and halt biodiversity loss</p>
          <img src={PovertyFirst} className={"life-first"}/>
          <img src={PovertySecond} className={"life-second"}/>
        </div>
        <div className={"small-box dark"}>
          <p className={"small-title"}>Peace, justice & strong institution</p>
          <p className={"small-txt"}>Promote peaceful and inclusive societies for sustainable development, provide access to justice for all and build effective, accountable and inclusive institutions at all levels</p>
          <img src={ZeroHunger} className={"peace-first"}/>
          <img src={ZeroHungerTop} className={"peace-top"}/>
        </div>
        <div className={"small-box light"}>
          <p className={"small-title"}>Partnerships for the goals</p>
          <p className={"small-txt"}>Strengthen the means of implementation and revitalize the global partnership for sustainable development</p>
          <img src={PovertyFirst} className={"partnership-first"}/>
          <img src={PovertySecond} className={"partnership-second"}/>
        </div>

        <div className={"small-box dark"}>
          <img src={ZeroHunger} className={"un-goals-first"}/>
          <img src={ZeroHungerTop} className={"un-goals-top"}/>
          <p className={"un-goals-title"}>THE UN SDG GOALS</p>
        </div>
      </div>
    </div>
  );
}