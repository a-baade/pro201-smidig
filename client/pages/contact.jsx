import React from "react";
import Message from "../pics/sign-up-pic/img_2.png";
import Phone from "../pics/icons/phone.png";
import Location from "../pics/icons/location.png";
import Pink from "../pics/contact-pics/pink-squiggle.png";
import Blue from "../pics/contact-pics/blue-squiggle.png";
import Person from "../pics/contact-pics/person-flowers.png";

export function Contact() {
    return (
        <div className={"login-container page-margin"}>
            <div className={"contact-sidebar"}>
                <h1 className={"sidebar-title"}>We'd love to <br/> hear from you</h1>
                <div>
                    <div className={"contact-info"}>
                        <img src={Phone} className={"contact-icon"} alt={"phoneIcon"}/>
                        <p>+47 999 99 999</p>
                    </div>
                    <div className={"contact-info"}>
                        <img src={Message} className={"contact-icon"} alt={"msgIcon"}/>
                        <p>meliora.impact@gmail.com</p>
                    </div>
                    <div className={"contact-info"}>
                        <img src={Location} className={"contact-icon"} alt={"locationIcon"}/>
                        <p>ch. groghsgate 12, 0186 Oslo Norway</p>
                    </div>
                </div>
                <div className={"contact-background"}>
                    <img src={Pink} className={"pink"} alt={"pinkSquig"}/>
                    <img src={Blue} className={"blue"} alt={"blueSquig"}/>
                    <img src={Person} className={"person"} alt={"personPic"}/>
                </div>
            </div>

            <div className={"contact-form"}>
                <h1>Get in touch with us</h1>
                <p>Fill out the form and our team will get back to you <br/> within 72 hours.</p>
                <div className={"name-field contact-input-field"}>
                    <label className={"login-txt contact-txt"}>Your name</label>
                    <input type= "text" placeholder={"Firstname Lastname"} className={"login-txt contact-input"}/>
                </div>
                <div className={"email-field contact-input-field"}>
                    <label className={"login-txt contact-txt"}>Email address</label>
                    <input type= "email" placeholder={"email@address.com"} className={"login-txt contact-input"}/>
                </div>
                <div className={"message-field contact-input-field"}>
                    <label className={"login-txt contact-txt"}>Your message</label>
                    <input type="textarea" placeholder={"Write your message"} className={"login-txt contact-textarea contact-input"}/>
                </div>
                <button className={"submit-btn sign-in-btn"}>Submit</button>
            </div>
        </div>
    );
}