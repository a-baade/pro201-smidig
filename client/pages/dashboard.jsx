import React from "react";
import { Chart2 } from "../components/charts";
import { Chart1 } from "../components/charts";


export function Dashboard({ user }) {
  return (
      <div className={"dashboard-page"}>
          <div className={"dashboard-sidebar"}>
              <div className={"dashboard-pfp-name"}>
                  <img src={user.google.picture} alt="Profile picture" className={"dashboard-pfp"}/>
                  <p>{user.google.name}</p>
              </div>
              <p>My donations</p>
              <p>My subscriptions</p>
          </div>
          <div className={"dashboard-main"}>
              <div className={"dashboard-top-row"}>
                <div className={"dashboard-info-box"}>
                    <p>23 <br/> donation</p>
                </div>
                  <div className={"dashboard-info-box"}>
                      <p>5 <br/> subscriptions</p>
                  </div>
                  <div className={"dashboard-info-box"}>
                      <p>750 000<br/> donated</p>
                  </div>
              </div>
              <div className={"dashboard-row-2"}>
                  <h2>Donation over time</h2>
                  <div className={"chart-wrap"}>
                      <div className={"chart2"}>
                          <Chart2/>
                      </div>
                      <div className={"chart2-sidebar"}>
                          <select name="months" id="months">
                              <option value="january">January</option>
                              <option value="february">February</option>
                              <option value="march">March</option>
                              <option value="april">April</option>
                          </select>
                          <div className={"chart2-circle"}>
                              <p>192 351 <br/> Avg per month</p>
                          </div>
                      </div>
                  </div>
              </div>
              <div className={"dashboard-row-3"}>
                  <h2>Donasjoner</h2>
                  <h4>sammenlignet med lignende bedrifter</h4>
                  <div className={"chart-wrap"}>
                      <div className={"chart1"}>
                          <Chart1/>
                      </div>
                      <div className={"chart1-sidebar"}>
                          <div className={"chart1-circle"}>
                              <p className={"percent-txt"}>25%</p>
                              <p className={"chart1-circle-txt"}>Bak gjennomsnittet</p>
                          </div>
                          <div className={"chart1-textbox"}>
                              <p>Basert p√• firma <br/> sin omsetning</p>
                          </div>
                      </div>
                  </div>
              </div>
          </div>
      </div>
  );
}