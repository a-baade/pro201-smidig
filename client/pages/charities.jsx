import React, { useContext, useEffect, useState } from "react";
import { Link } from "react-router-dom";
import { ApiContext } from "../apiContext";
import { useLoading } from "../useLoading";
import { fetchJSON } from "../lib/fetchJSON";
import LoadingWheel from "../components/loadingSpinner";

function CharityCard({
  charity: { _id, name, description, bgImage, charityLogo },
}) {
  // Charity card with button to donate or charity page
  return (
    <div className={"charity-card custom-shadow"}>
      <div
        className={"charity-background"}
        style={{ backgroundImage: `url( data:image/png;base64${bgImage} )` }}
      >
        <img
          src={`data:image/png;base64${charityLogo}`}
          className={"charity-img custom-shadow"}
        />
      </div>

      <div className={"charity-card-content"}>
        <h3>{name}</h3>
        <p>{description}</p>
        <div className={"card-buttons"}>
          <Link to={{ pathname: "/charities/charity/id?id=" + _id }}>
            <button className="card-button custom-shadow-button">
              Read more
            </button>
          </Link>
          <Link to={{ pathname: "/charities/donate/id?id=" + _id }}>
            <button className="card-button custom-shadow-button">Donate</button>
          </Link>
        </div>
      </div>
    </div>
  );
}

export function Charities() {
  const { allCharities } = useContext(ApiContext);
  const { category, setCategory } = useState("");
  const { categoryQuery, setCategoryQuery } = useState("");
  const [search, setSearch] = useState("");
  const [data, setData] = useState([]);
  const { loading, error } = useLoading(
    async () => await allCharities({ category }),
    [category]
  );

  useEffect(async () => {
    if (search === "") {
      await fetchJSON("/api/charities").then((jsonData) => {
        setData(jsonData);
      });
    } else {
      await fetchJSON(`/api/charities/search/?name=${search}`).then(
        (jsonData) => {
          setData(jsonData);
        }
      );
    }
  }, [search]);

  function handleSubmitQuery(e) {
    e.preventDefault();
    setCategory(categoryQuery);
  }

  async function handleSearch(event) {
    await setSearch(event);
  }

  if (loading) {
    return <LoadingWheel />;
  }

  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  // Add category filter
  return (
    <div className={"page-margin"}>
      <div className="container-charities">
        <div className="search-charities">
          <input
            className="search-input-charities"
            placeholder="Search Charities"
            value={search}
            onChange={(e) => handleSearch(e.target.value)}
          />
          {search && (
            <div className={"charity-results"}>Results for {search}</div>
          )}
        </div>
        <div></div>

        {/*<form onSubmit={handleSubmitQuery}>*/}
        {/*    <button  id={"category-query"}*/}
        {/*             value={categoryQuery}*/}
        {/*             onChange={(e) => setCategoryQuery(e.target.value)}>Water</button>*/}
        {/*    <button  id={"category-query"}*/}
        {/*             value={categoryQuery}*/}
        {/*             onChange={(e) => setCategoryQuery(e.target.value)}>Knowledge</button>*/}
        {/*</form>*/}
      </div>
      <div>
        <ul className={"charity-cards-container"}>
          {data?.map((charity, index) => (
            <CharityCard key={index} charity={charity} />
          ))}
        </ul>
      </div>
    </div>
  );
}
