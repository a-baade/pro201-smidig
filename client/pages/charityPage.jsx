import React, { useContext } from "react";
import { Link, useSearchParams} from "react-router-dom";
import { ApiContext } from "../apiContext";
import { useLoading } from "../useLoading";
import Arrow from "../pics/icons/arrow.png";
import NewsImg1 from "../pics/bg-pictures/img_3.png";
import NewsImg2 from "../pics/bg-pictures/img_4.png";
import NewsImg3 from "../pics/bg-pictures/img_5.jpg";
import Facebook from "../pics/icons/facebook.png";
import Instagram from "../pics/icons/instagram.png";
import Snapchat from "../pics/icons/snapchat.png";
import Tiktok from "../pics/icons/tiktok.png";
import WWW from "../pics/icons/world-wide-web.png";
import LoadingWheel from "../components/loadingSpinner";

function Charity({charity:
    { _id, name, description, bgImage, charityLogo }
}) {

  // Charity page
  return (
    <div>
      <Link to={"/charities"}>
        <div className={"back-button custom-shadow-button"}>
          <img src={Arrow} />
        </div>
      </Link>
      <img
        src={`data:image/png;base64${bgImage}`}
        className={"charity-page-background"}
      />
      <div className={"charity-page-content"}>
        <div className={"charity-page-description"}>
          <div className={"charity-description-top"}>
            <img
              src={`data:image/png;base64${charityLogo}`}
              className={"charity-description-logo"}
            />
            <h2>{name}</h2>
          </div>
          <div className={"charity-description-text"}>
            <p>{description}</p>
          </div>
        </div>
        <div className={"charity-page-side"}>
          <Link to={{ pathname: "/charities/donate/id?id=" + _id }}>
            <button className={"charity-page-donate-button custom-shadow-button"}>Donate</button>
          </Link>
          <a
            href={"https://www.google.com/"}
            className={"charity-social-links"}
          >
            <img src={WWW} /> examplecharity.com
          </a>
          <a
            href={"https://www.facebook.com/"}
            className={"charity-social-links"}
          >
            <img src={Facebook} /> Example Charity
          </a>
          <a
            href={"https://www.instagram.com/"}
            className={"charity-social-links"}
          >
            <img src={Instagram} /> @examplecharity
          </a>
          <a
            href={"https://www.tiktok.com/"}
            className={"charity-social-links"}
          >
            <img src={Tiktok} /> @examplecharity
          </a>
          <a
            href={"https://www.snapchat.com/"}
            className={"charity-social-links"}
          >
            <img src={Snapchat} /> examplecharity
          </a>
        </div>
      </div>
      <div className={"newsfeed-container custom-shadow"}>
        <h1>NEWSFEED</h1>
        <div className={"newsfeed-items-container"}>
          <div className={"newsfeed-item custom-shadow"}>
            <img
                src={NewsImg1}
                className={"newsfeed-image"}
                alt={"Afghan children in school"}
            />
            <p>
              Afghan refugee children enrolling in norwegian nurseries 2 weeks
              after coming to norway. Prime minister hopes that these efforts
              will help stabilize the everyday life of the children and that
              they will integrate quicker.
            </p>
            <p className={"read-more"}>Read more</p>
          </div>
          <div className={"newsfeed-item custom-shadow"}>
            <img
                src={NewsImg2}
                className={"newsfeed-image"}
                alt={"Deers in nature"}
            />
            <p>
              Hunting season has ended, yet people are still hunting animals
              illegally. Protecting nature is our responsibility, and make sure
              that the next generation can enjoy wilderness like us.
            </p>
            <p className={"read-more"}>Read more</p>
          </div>
          <div className={"newsfeed-item custom-shadow"}>
            <img
                src={NewsImg3}
                className={"newsfeed-image"}
                alt={"Akeem in karate training"}
            />
            <p>
              Thanks to the support of charities, children like Akeem are able
              to partake in after school activities. Children like Akeem would
              not be able to afford ...
            </p>
            <p className={"read-more"}>Read more</p>
          </div>
        </div>
      </div>
    </div>
  );
}

export default function CharityPage() {
  let [searchParams, setSearchParams] = useSearchParams();
  const id = searchParams.get("id");

  const {oneCharity} = useContext(ApiContext);
  const {loading, error, data} = useLoading(
    async () => await oneCharity(id),
    []
  );

  if (loading) {
    return <LoadingWheel/>;;
  }

  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  return (
    <div className={"page-margin"}>
      <ul className={"charity-page-container"}>
        {data.map((charity, index) => (
          <Charity key={index} charity={charity} />
        ))}
      </ul>
    </div>
  );
}
