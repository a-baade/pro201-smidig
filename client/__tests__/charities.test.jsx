import React from "react";
import ReactDOM from "react-dom";
import { Charities } from "../pages/charities";
import { act, Simulate } from "react-dom/test-utils";
import { ApiContext } from "../apiContext";
import { BrowserRouter } from "react-router-dom";
import CharityPage from "../pages/charityPage";

describe("Charities component", () => {
  it("should show loading screen", function () {
    const divElement = document.createElement("div");
    ReactDOM.render(<Charities />, divElement);
    expect(divElement.innerHTML).toMatchSnapshot();
  });

  it("should show error message", async function () {
    const divElement = document.createElement("div");
    await act(async () => {
      const allCharities = () => {
        throw new Error("Failed to load");
      };
      ReactDOM.render(
        <ApiContext.Provider value={{ allCharities }}>
          <Charities />
        </ApiContext.Provider>,
        divElement
      );
    });
    expect(divElement.querySelector("#error-text").innerHTML).toEqual(
      "Error: Failed to load"
    );
    expect(divElement.innerHTML).toMatchSnapshot();
  });

  it("should show all charities", async function () {
    const charities = [
      { name: "Charity 1", description: "Test description 1" },
      { name: "Charity 2", description: "Test description 2" },
      { name: "Charity 3", description: "Test description 3" },
    ];
    const divElement = document.createElement("div");
    await act(async () => {
      ReactDOM.render(
        <BrowserRouter>
          <ApiContext.Provider value={{ allCharities: () => charities }}>
            <Charities />
          </ApiContext.Provider>
        </BrowserRouter>,
        divElement
      );
    });
    expect(
      Array.from(divElement.querySelectorAll("h3")).map((e) => e.innerHTML)
    ).toEqual(["Charity 1", "Charity 2", "Charity 3"]);
    expect(
      Array.from(divElement.querySelectorAll("p")).map((e) => e.innerHTML)
    ).toEqual([
      "Test description 1",
      "Test description 2",
      "Test description 3",
    ]);
    expect(divElement.innerHTML).toMatchSnapshot();
  });

  it("should show charity page", async function () {
    const charities = [
      { id: 1, name: "Charity 1", description: "Test description 1" },
      { id: 2, name: "Charity 2", description: "Test description 2" },
      { id: 3, name: "Charity 3", description: "Test description 3" },
    ];

    const divElement = document.createElement("div");
    await act(async () => {
      ReactDOM.render(
        <BrowserRouter>
          <ApiContext.Provider value={{ oneCharity: () => charities }}>
            <CharityPage />
          </ApiContext.Provider>
        </BrowserRouter>,
        divElement
      );

      expect(divElement.innerHTML).toMatchSnapshot();
    });
  });
});
