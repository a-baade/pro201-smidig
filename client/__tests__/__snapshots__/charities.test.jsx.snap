// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Charities component should show all charities 1`] = `"<div class=\\"page-margin\\"><div><ul class=\\"charity-parent-cont\\"><div class=\\"charity-card\\"><div class=\\"charity-background\\"><img src=\\"data:image/png;base64undefined\\" class=\\"charity-img\\"></div><div class=\\"charity-card-content\\"><h3>Charity 1</h3><p>Test description 1</p><div class=\\"card-buttons\\"><a href=\\"/charities/charity/id?id=undefined\\"><button class=\\"card-btn-left card-btn\\">Read more</button></a><a href=\\"/charities/donate/id?id=undefined\\"><button class=\\"card-btn-right card-btn\\">Donate</button></a></div></div></div><div class=\\"charity-card\\"><div class=\\"charity-background\\"><img src=\\"data:image/png;base64undefined\\" class=\\"charity-img\\"></div><div class=\\"charity-card-content\\"><h3>Charity 2</h3><p>Test description 2</p><div class=\\"card-buttons\\"><a href=\\"/charities/charity/id?id=undefined\\"><button class=\\"card-btn-left card-btn\\">Read more</button></a><a href=\\"/charities/donate/id?id=undefined\\"><button class=\\"card-btn-right card-btn\\">Donate</button></a></div></div></div><div class=\\"charity-card\\"><div class=\\"charity-background\\"><img src=\\"data:image/png;base64undefined\\" class=\\"charity-img\\"></div><div class=\\"charity-card-content\\"><h3>Charity 3</h3><p>Test description 3</p><div class=\\"card-buttons\\"><a href=\\"/charities/charity/id?id=undefined\\"><button class=\\"card-btn-left card-btn\\">Read more</button></a><a href=\\"/charities/donate/id?id=undefined\\"><button class=\\"card-btn-right card-btn\\">Donate</button></a></div></div></div></ul></div></div>"`;

exports[`Charities component should show charity page 1`] = `"<div>Loading...</div>"`;

exports[`Charities component should show error message 1`] = `"<div><h1>Error</h1><div id=\\"error-text\\">Error: Failed to load</div></div>"`;

exports[`Charities component should show loading screen 1`] = `"<div>Loading...</div>"`;
